File: attocode/src/tricks/failure-evidence.ts
/**
 * Trick S: Failure Evidence Preservation
 *
 * Preserves failed actions, error traces, and unsuccessful attempts in context
 * so the model can learn from mistakes and avoid repeating them.
 *
 * Problem: When errors are silently discarded or minimally logged, the model
 * may repeat the same mistakes. Without evidence of what failed and why,
 * the agent loops endlessly.
 *
 * Solution: Explicitly preserve failure evidence with structured metadata.
 * This includes error messages, stack traces, what was attempted, and why
 * it failed - enabling the model to learn and adapt.
 *
 * @example
 * ```typescript
 * import { createFailureTracker, formatFailureContext } from './failure-evidence';
 *
 * const tracker = createFailureTracker({
 *   maxFailures: 20,
 *   preserveStackTraces: true,
 *   categorizeErrors: true,
 * });
 *
 * // Record a failure
 * tracker.recordFailure({
 *   action: 'read_file',
 *   args: { path: '/etc/passwd' },
 *   error: 'Permission denied',
 *   category: 'permission',
 * });
 *
 * // Get context for LLM
 * const failureContext = tracker.getFailureContext();
 * ```
 */

import { logger } from '../integrations/utilities/logger.js';

// =============================================================================
// TYPES
// =============================================================================

/**
 * Categories of failures for pattern detection.
 */
export type FailureCategory =
  | 'permission'     // Access/permission denied
  | 'not_found'      // File/resource not found
  | 'syntax'         // Syntax/parsing errors
  | 'type'           // Type errors
  | 'runtime'        // Runtime exceptions
  | 'network'        // Network/connection errors
  | 'timeout'        // Timeout errors
  | 'validation'     // Input validation errors
  | 'logic'          // Logic/assertion errors
  | 'resource'       // Resource exhaustion (memory, disk)
  | 'unknown';       // Uncategorized

/**
 * A recorded failure.
 */
export interface Failure {
  /** Unique ID */
  id: string;

  /** When the failure occurred */
  timestamp: string;

  /** What action/tool was attempted */
  action: string;

  /** Arguments passed to the action */
  args?: Record<string, unknown>;

  /** Error message */
  error: string;

  /** Full stack trace (if available) */
  stackTrace?: string;

  /** Error category */
  category: FailureCategory;

  /** Iteration/turn when this occurred */
  iteration?: number;

  /** What was the intent/goal */
  intent?: string;

  /** Suggested fix or workaround (from error analysis) */
  suggestion?: string;

  /** Whether this failure has been addressed */
  resolved: boolean;

  /** Number of times similar failure occurred */
  repeatCount: number;
}

/**
 * Configuration for failure tracking.
 */
export interface FailureTrackerConfig {
  /** Maximum failures to keep in memory */
  maxFailures?: number;

  /** Whether to preserve full stack traces */
  preserveStackTraces?: boolean;

  /** Whether to auto-categorize errors */
  categorizeErrors?: boolean;

  /** Whether to detect repeated failures */
  detectRepeats?: boolean;

  /** Threshold for warning about repeated failures */
  repeatWarningThreshold?: number;
}

/**
 * Failure input for recording.
 */
export interface FailureInput {
  /** What action/tool was attempted */
  action: string;

  /** Arguments passed */
  args?: Record<string, unknown>;

  /** Error message or Error object */
  error: string | Error;

  /** Error category (auto-detected if not provided) */
  category?: FailureCategory;

  /** Current iteration */
  iteration?: number;

  /** What was the goal */
  intent?: string;
}

/**
 * Pattern detected in failures.
 */
export interface FailurePattern {
  /** Pattern type */
  type: 'repeated_action' | 'repeated_error' | 'category_cluster' | 'escalating';

  /** Description of the pattern */
  description: string;

  /** Related failure IDs */
  failureIds: string[];

  /** Confidence (0-1) */
  confidence: number;

  /** Suggested action */
  suggestion: string;
}

/**
 * Events emitted by failure tracker.
 */
export type FailureEvent =
  | { type: 'failure.recorded'; failure: Failure }
  | { type: 'failure.repeated'; failure: Failure; count: number }
  | { type: 'pattern.detected'; pattern: FailurePattern }
  | { type: 'failure.resolved'; failureId: string }
  | { type: 'failure.evicted'; failure: Failure; reason: string };

export type FailureEventListener = (event: FailureEvent) => void;

// =============================================================================
// ERROR CATEGORIZATION
// =============================================================================

/**
 * Auto-categorize an error based on its message.
 */
export function categorizeError(error: string): FailureCategory {
  const lowerError = error.toLowerCase();

  // Permission errors
  if (
    lowerError.includes('permission denied') ||
    lowerError.includes('access denied') ||
    lowerError.includes('not permitted') ||
    lowerError.includes('eacces') ||
    lowerError.includes('unauthorized')
  ) {
    return 'permission';
  }

  // Not found errors
  if (
    lowerError.includes('not found') ||
    lowerError.includes('no such file') ||
    lowerError.includes('enoent') ||
    lowerError.includes('does not exist') ||
    lowerError.includes('404')
  ) {
    return 'not_found';
  }

  // Syntax errors
  if (
    lowerError.includes('syntax error') ||
    lowerError.includes('unexpected token') ||
    lowerError.includes('parse error') ||
    lowerError.includes('invalid json')
  ) {
    return 'syntax';
  }

  // Type errors
  if (
    lowerError.includes('type error') ||
    lowerError.includes('typeerror') ||
    lowerError.includes('is not a function') ||
    lowerError.includes('undefined is not') ||
    lowerError.includes('cannot read propert')
  ) {
    return 'type';
  }

  // Network errors
  if (
    lowerError.includes('network') ||
    lowerError.includes('connection') ||
    lowerError.includes('econnrefused') ||
    lowerError.includes('socket') ||
    lowerError.includes('dns') ||
    lowerError.includes('fetch failed')
  ) {
    return 'network';
  }

  // Timeout errors
  if (
    lowerError.includes('timeout') ||
    lowerError.includes('timed out') ||
    lowerError.includes('etimedout')
  ) {
    return 'timeout';
  }

  // Validation errors
  if (
    lowerError.includes('validation') ||
    lowerError.includes('invalid') ||
    lowerError.includes('required') ||
    lowerError.includes('must be')
  ) {
    return 'validation';
  }

  // Resource errors
  if (
    lowerError.includes('out of memory') ||
    lowerError.includes('disk full') ||
    lowerError.includes('enomem') ||
    lowerError.includes('enospc') ||
    lowerError.includes('quota')
  ) {
    return 'resource';
  }

  // Logic errors
  if (
    lowerError.includes('assertion') ||
    lowerError.includes('invariant') ||
    lowerError.includes('expect')
  ) {
    return 'logic';
  }

  return 'unknown';
}

/**
 * Generate a suggestion based on failure category.
 */
export function generateSuggestion(failure: Failure): string {
  switch (failure.category) {
    case 'permission':
      return `Check file/directory permissions. The action "${failure.action}" may need elevated privileges or the path may be restricted.`;

    case 'not_found':
      return `Verify the resource exists before accessing. Use list_directory or check for typos in the path.`;

    case 'syntax':
      return `Review the syntax. Check for missing brackets, quotes, or invalid characters. Validate JSON/code before submission.`;

    case 'type':
      return `Check variable types and ensure proper null/undefined handling. The value may be a different type than expected.`;

    case 'network':
      return `Check network connectivity. The service may be down, or there may be firewall/DNS issues.`;

    case 'timeout':
      return `The operation took too long. Consider breaking it into smaller operations or increasing timeout.`;

    case 'validation':
      return `Verify input matches expected format/constraints. Check required fields and value ranges.`;

    case 'resource':
      return `System resources are constrained. Consider cleanup, smaller batch sizes, or freeing memory/disk.`;

    case 'logic':
      return `An assumption was violated. Review the logic and check preconditions.`;

    default:
      return `Analyze the error message for specific guidance.`;
  }
}

// =============================================================================
// FAILURE TRACKER
// =============================================================================

/**
 * Tracks failures with pattern detection.
 */
export class FailureTracker {
  private config: Required<FailureTrackerConfig>;
  private failures: Failure[] = [];
  private actionHistory: Map<string, string[]> = new Map(); // action -> failure IDs
  private listeners: FailureEventListener[] = [];

  constructor(config: FailureTrackerConfig = {}) {
    this.config = {
      maxFailures: config.maxFailures ?? 50,
      preserveStackTraces: config.preserveStackTraces ?? true,
      categorizeErrors: config.categorizeErrors ?? true,
      detectRepeats: config.detectRepeats ?? true,
      repeatWarningThreshold: config.repeatWarningThreshold ?? 3,
    };
  }

  /**
   * Record a failure.
   */
  recordFailure(input: FailureInput): Failure {
    const errorString = input.error instanceof Error
      ? input.error.message
      : input.error;

    const stackTrace = input.error instanceof Error
      ? input.error.stack
      : undefined;

    // Auto-categorize if needed
    const category = input.category ??
      (this.config.categorizeErrors ? categorizeError(errorString) : 'unknown');

    // Check for repeats
    let repeatCount = 1;
    if (this.config.detectRepeats) {
      repeatCount = this.countSimilarFailures(input.action, errorString);
    }

    const failure: Failure = {
      id: `fail-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
      timestamp: new Date().toISOString(),
      action: input.action,
      args: input.args,
      error: errorString,
      stackTrace: this.config.preserveStackTraces ? stackTrace : undefined,
      category,
      iteration: input.iteration,
      intent: input.intent,
      suggestion: '',
      resolved: false,
      repeatCount,
    };

    // Generate suggestion
    failure.suggestion = generateSuggestion(failure);

    // Store
    this.failures.push(failure);

    // Track by action
    const actionFailures = this.actionHistory.get(input.action) || [];
    actionFailures.push(failure.id);
    this.actionHistory.set(input.action, actionFailures);

    // Enforce max failures with eviction warning
    if (this.failures.length > this.config.maxFailures) {
      const evicted = this.failures.shift()!;

      // Log warning about eviction
      logger.warn(`[FailureTracker] Evicted failure due to limit (${this.config.maxFailures}): ${evicted.action} - ${evicted.error.slice(0, 50)}`);

      // Emit eviction event for observability
      this.emit({
        type: 'failure.evicted',
        failure: evicted,
        reason: `Max failures limit (${this.config.maxFailures}) exceeded`,
      });

      // Clean up action history
      for (const [_action, ids] of this.actionHistory) {
        const idx = ids.indexOf(evicted.id);
        if (idx >= 0) ids.splice(idx, 1);
      }
    }

    // Emit events
    this.emit({ type: 'failure.recorded', failure });

    if (repeatCount >= this.config.repeatWarningThreshold) {
      this.emit({ type: 'failure.repeated', failure, count: repeatCount });
    }

    // Check for patterns
    this.detectPatterns();

    return failure;
  }

  /**
   * Mark a failure as resolved.
   */
  resolveFailure(failureId: string): boolean {
    const failure = this.failures.find(f => f.id === failureId);
    if (failure) {
      failure.resolved = true;
      this.emit({ type: 'failure.resolved', failureId });
      return true;
    }
    return false;
  }

  /**
   * Get all unresolved failures.
   */
  getUnresolvedFailures(): Failure[] {
    return this.failures.filter(f => !f.resolved);
  }

  /**
   * Get failures by category.
   */
  getFailuresByCategory(category: FailureCategory): Failure[] {
    return this.failures.filter(f => f.category === category);
  }

  /**
   * Get failures by action.
   */
  getFailuresByAction(action: string): Failure[] {
    const ids = this.actionHistory.get(action) || [];
    return this.failures.filter(f => ids.includes(f.id));
  }

  /**
   * Get recent failures (last N).
   */
  getRecentFailures(count: number = 10): Failure[] {
    return this.failures.slice(-count);
  }

  /**
   * Get failure context formatted for LLM inclusion.
   */
  getFailureContext(options: {
    maxFailures?: number;
    includeResolved?: boolean;
    includeStackTraces?: boolean;
  } = {}): string {
    const {
      maxFailures = 10,
      includeResolved = false,
      includeStackTraces = false,
    } = options;

    let failures = includeResolved
      ? this.failures
      : this.failures.filter(f => !f.resolved);

    failures = failures.slice(-maxFailures);

    if (failures.length === 0) {
      return '';
    }

    return formatFailureContext(failures, { includeStackTraces });
  }

  /**
   * Check if an action has failed recently.
   */
  hasRecentFailure(action: string, withinMs: number = 60000): boolean {
    const cutoff = Date.now() - withinMs;
    return this.failures.some(
      f => f.action === action && new Date(f.timestamp).getTime() > cutoff
    );
  }

  /**
   * Get failure statistics.
   */
  getStats(): {
    total: number;
    unresolved: number;
    byCategory: Record<FailureCategory, number>;
    mostFailedActions: Array<{ action: string; count: number }>;
  } {
    const byCategory: Record<FailureCategory, number> = {
      permission: 0,
      not_found: 0,
      syntax: 0,
      type: 0,
      runtime: 0,
      network: 0,
      timeout: 0,
      validation: 0,
      logic: 0,
      resource: 0,
      unknown: 0,
    };

    for (const f of this.failures) {
      byCategory[f.category]++;
    }

    const actionCounts = new Map<string, number>();
    for (const f of this.failures) {
      actionCounts.set(f.action, (actionCounts.get(f.action) || 0) + 1);
    }

    const mostFailedActions = Array.from(actionCounts.entries())
      .map(([action, count]) => ({ action, count }))
      .sort((a, b) => b.count - a.count)
      .slice(0, 5);

    return {
      total: this.failures.length,
      unresolved: this.failures.filter(f => !f.resolved).length,
      byCategory,
      mostFailedActions,
    };
  }

  /**
   * Clear all failures.
   */
  clear(): void {
    this.failures = [];
    this.actionHistory.clear();
  }

  /**
   * Subscribe to events.
   */
  on(listener: FailureEventListener): () => void {
    this.listeners.push(listener);
    return () => {
      const idx = this.listeners.indexOf(listener);
      if (idx >= 0) this.listeners.splice(idx, 1);
    };
  }

  // Internal methods

  private countSimilarFailures(action: string, error: string): number {
    const normalizedError = error.toLowerCase().slice(0, 50);
    return this.failures.filter(
      f => f.action === action && f.error.toLowerCase().startsWith(normalizedError)
    ).length + 1;
  }

  private detectPatterns(): void {
    // Detect repeated action failures
    for (const [action, ids] of this.actionHistory) {
      if (ids.length >= 3) {
        const recent = ids.slice(-3);
        const recentFailures = this.failures.filter(f => recent.includes(f.id));

        if (recentFailures.every(f => !f.resolved)) {
          this.emit({
            type: 'pattern.detected',
            pattern: {
              type: 'repeated_action',
              description: `Action "${action}" has failed ${ids.length} times`,
              failureIds: recent,
              confidence: Math.min(0.9, 0.3 + ids.length * 0.1),
              suggestion: `Consider an alternative approach. "${action}" is consistently failing.`,
            },
          });
        }
      }
    }

    // Detect category clusters
    const recentFailures = this.failures.slice(-10);
    const categoryCounts = new Map<FailureCategory, number>();

    for (const f of recentFailures) {
      categoryCounts.set(f.category, (categoryCounts.get(f.category) || 0) + 1);
    }

    for (const [category, count] of categoryCounts) {
      if (count >= 5) {
        this.emit({
          type: 'pattern.detected',
          pattern: {
            type: 'category_cluster',
            description: `Multiple ${category} errors (${count} of last 10)`,
            failureIds: recentFailures.filter(f => f.category === category).map(f => f.id),
            confidence: count / 10,
            suggestion: `Address the underlying ${category} issue before continuing.`,
          },
        });
      }
    }
  }

  private emit(event: FailureEvent): void {
    for (const listener of this.listeners) {
      try {
        listener(event);
      } catch {
        // Ignore listener errors
      }
    }
  }
}

// =============================================================================
// FACTORY FUNCTIONS
// =============================================================================

/**
 * Create a failure tracker.
 *
 * @example
 * ```typescript
 * const tracker = createFailureTracker({
 *   maxFailures: 30,
 *   preserveStackTraces: true,
 *   detectRepeats: true,
 * });
 *
 * // Record failures during agent loop
 * try {
 *   await tool.execute(args);
 * } catch (error) {
 *   tracker.recordFailure({
 *     action: tool.name,
 *     args,
 *     error,
 *     iteration: currentIteration,
 *   });
 * }
 *
 * // Include failure context in prompts
 * const failureContext = tracker.getFailureContext();
 * if (failureContext) {
 *   messages.push({
 *     role: 'system',
 *     content: failureContext,
 *   });
 * }
 * ```
 */
export function createFailureTracker(
  config: FailureTrackerConfig = {}
): FailureTracker {
  return new FailureTracker(config);
}

// =============================================================================
// UTILITIES
// =============================================================================

/**
 * Format failures as context for LLM.
 */
export function formatFailureContext(
  failures: Failure[],
  options: { includeStackTraces?: boolean } = {}
): string {
  if (failures.length === 0) return '';

  const lines = [
    '[Previous Failures - Learn from these to avoid repeating mistakes]',
    '',
  ];

  for (const f of failures) {
    lines.push(`**${f.action}** (${f.category}): ${f.error}`);

    if (f.args && Object.keys(f.args).length > 0) {
      lines.push(`  Args: ${JSON.stringify(f.args)}`);
    }

    if (f.suggestion) {
      lines.push(`  → ${f.suggestion}`);
    }

    if (options.includeStackTraces && f.stackTrace) {
      const shortStack = f.stackTrace.split('\n').slice(0, 3).join('\n');
      lines.push(`  Stack: ${shortStack}`);
    }

    lines.push('');
  }

  return lines.join('\n');
}

/**
 * Create a warning message for repeated failures.
 */
export function createRepeatWarning(
  action: string,
  count: number,
  suggestion?: string
): string {
  let warning = `⚠️ Action "${action}" has failed ${count} times.`;

  if (count >= 5) {
    warning += ' Consider a completely different approach.';
  } else if (count >= 3) {
    warning += ' Review previous errors before retrying.';
  }

  if (suggestion) {
    warning += `\n💡 Suggestion: ${suggestion}`;
  }

  return warning;
}

/**
 * Extract actionable insights from failures.
 */
export function extractInsights(failures: Failure[]): string[] {
  const insights: string[] = [];

  // Group by category
  const byCategory = new Map<FailureCategory, Failure[]>();
  for (const f of failures) {
    const group = byCategory.get(f.category) || [];
    group.push(f);
    byCategory.set(f.category, group);
  }

  // Generate insights
  if (byCategory.has('permission') && byCategory.get('permission')!.length >= 2) {
    insights.push('Multiple permission errors - check if running with sufficient privileges');
  }

  if (byCategory.has('not_found') && byCategory.get('not_found')!.length >= 2) {
    insights.push('Multiple not-found errors - verify paths and use list_directory first');
  }

  if (byCategory.has('syntax') && byCategory.get('syntax')!.length >= 2) {
    insights.push('Multiple syntax errors - validate code/JSON before execution');
  }

  if (byCategory.has('network') && byCategory.get('network')!.length >= 1) {
    insights.push('Network issues detected - check connectivity and service availability');
  }

  if (byCategory.has('timeout') && byCategory.get('timeout')!.length >= 1) {
    insights.push('Timeout detected - consider breaking operations into smaller chunks');
  }

  // Action-specific insights
  const actionCounts = new Map<string, number>();
  for (const f of failures) {
    actionCounts.set(f.action, (actionCounts.get(f.action) || 0) + 1);
  }

  for (const [action, count] of actionCounts) {
    if (count >= 3) {
      insights.push(`"${action}" failed ${count} times - try an alternative tool/approach`);
    }
  }

  return insights;
}

/**
 * Format failure stats for display.
 */
export function formatFailureStats(stats: ReturnType<FailureTracker['getStats']>): string {
  const lines = [
    `Failure Statistics:`,
    `  Total: ${stats.total} (${stats.unresolved} unresolved)`,
    '',
    '  By Category:',
  ];

  for (const [category, count] of Object.entries(stats.byCategory)) {
    if (count > 0) {
      lines.push(`    ${category}: ${count}`);
    }
  }

  if (stats.mostFailedActions.length > 0) {
    lines.push('');
    lines.push('  Most Failed Actions:');
    for (const { action, count } of stats.mostFailedActions) {
      lines.push(`    ${action}: ${count}`);
    }
  }

  return lines.join('\n');
}

File: attocode/attocode_py/src/attocode/tricks/failure_evidence.py
"""Failure evidence preservation (Trick S).

Tracks tool failures, detects patterns, generates suggestions,
and formats failure context for LLM injection.
"""

from __future__ import annotations

import re
import time
import uuid
from dataclasses import dataclass, field
from enum import StrEnum
from typing import Any, Callable


class FailureCategory(StrEnum):
    """Categories of failures."""

    PERMISSION = "permission"
    NOT_FOUND = "not_found"
    SYNTAX = "syntax"
    TYPE = "type"
    RUNTIME = "runtime"
    NETWORK = "network"
    TIMEOUT = "timeout"
    VALIDATION = "validation"
    LOGIC = "logic"
    RESOURCE = "resource"
    UNKNOWN = "unknown"


@dataclass
class Failure:
    """A recorded failure."""

    id: str
    timestamp: float
    action: str
    error: str
    category: FailureCategory = FailureCategory.UNKNOWN
    args: dict[str, Any] | None = None
    stack_trace: str | None = None
    iteration: int | None = None
    intent: str | None = None
    suggestion: str | None = None
    resolved: bool = False
    repeat_count: int = 1


@dataclass
class FailureInput:
    """Input for recording a failure."""

    action: str
    error: str | Exception
    args: dict[str, Any] | None = None
    category: FailureCategory | None = None
    iteration: int | None = None
    intent: str | None = None


@dataclass
class FailurePattern:
    """A detected failure pattern."""

    type: str  # repeated_action, repeated_error, category_cluster, escalating
    description: str
    failure_ids: list[str]
    confidence: float = 0.5
    suggestion: str | None = None


@dataclass
class FailureTrackerConfig:
    """Configuration for failure tracking."""

    max_failures: int = 50
    preserve_stack_traces: bool = True
    categorize_errors: bool = True
    detect_repeats: bool = True
    repeat_warning_threshold: int = 3


FailureEventListener = Callable[[str, dict[str, Any]], None]

# Categorization patterns
_CATEGORY_PATTERNS: list[tuple[re.Pattern[str], FailureCategory]] = [
    (re.compile(r"EACCES|permission denied|PermissionError|forbidden", re.IGNORECASE), FailureCategory.PERMISSION),
    (re.compile(r"ENOENT|not found|FileNotFoundError|no such file", re.IGNORECASE), FailureCategory.NOT_FOUND),
    (re.compile(r"syntax error|SyntaxError|unexpected token|parse error", re.IGNORECASE), FailureCategory.SYNTAX),
    (re.compile(r"type error|TypeError|invalid type|wrong type", re.IGNORECASE), FailureCategory.TYPE),
    (re.compile(r"network|ECONNREFUSED|EHOSTUNREACH|DNS|ConnectionError", re.IGNORECASE), FailureCategory.NETWORK),
    (re.compile(r"timeout|ETIMEDOUT|timed? ?out|TimeoutError", re.IGNORECASE), FailureCategory.TIMEOUT),
    (re.compile(r"validation|invalid|ValueError|required field", re.IGNORECASE), FailureCategory.VALIDATION),
    (re.compile(r"out of memory|ENOMEM|MemoryError|disk full", re.IGNORECASE), FailureCategory.RESOURCE),
    (re.compile(r"RuntimeError|runtime|assertion|AssertionError", re.IGNORECASE), FailureCategory.RUNTIME),
]


class FailureTracker:
    """Tracks and analyzes agent failures.

    Records failures, auto-categorizes them, detects repeat
    patterns, and generates actionable suggestions.
    """

    def __init__(self, config: FailureTrackerConfig | None = None) -> None:
        self._config = config or FailureTrackerConfig()
        self._failures: list[Failure] = []
        self._listeners: list[FailureEventListener] = []

    def record_failure(self, input: FailureInput) -> Failure:
        """Record a new failure."""
        # Extract error string and stack trace
        if isinstance(input.error, Exception):
            error_str = str(input.error)
            stack_trace = None
            if self._config.preserve_stack_traces and hasattr(input.error, "__traceback__"):
                import traceback
                stack_trace = "".join(traceback.format_exception(type(input.error), input.error, input.error.__traceback__))
        else:
            error_str = input.error
            stack_trace = None

        # Categorize
        category = input.category or (
            categorize_error(error_str) if self._config.categorize_errors
            else FailureCategory.UNKNOWN
        )

        # Detect repeats
        repeat_count = 1
        if self._config.detect_repeats:
            prefix = error_str[:50].lower()
            repeat_count = 1 + sum(
                1 for f in self._failures
                if f.action == input.action
                and f.error[:50].lower() == prefix
                and not f.resolved
            )

        failure = Failure(
            id=f"fail-{uuid.uuid4().hex[:8]}",
            timestamp=time.time(),
            action=input.action,
            error=error_str,
            category=category,
            args=input.args,
            stack_trace=stack_trace,
            iteration=input.iteration,
            intent=input.intent,
            repeat_count=repeat_count,
            suggestion=None,
        )

        # Generate suggestion
        failure.suggestion = generate_suggestion(failure)

        # Enforce max
        if len(self._failures) >= self._config.max_failures:
            evicted = self._failures.pop(0)
            self._emit("failure.evicted", {"failure_id": evicted.id})

        self._failures.append(failure)
        self._emit("failure.recorded", {"failure_id": failure.id, "category": category})

        if repeat_count >= self._config.repeat_warning_threshold:
            self._emit("failure.repeated", {
                "failure_id": failure.id,
                "action": failure.action,
                "count": repeat_count,
            })

        # Detect patterns
        self._detect_patterns()

        return failure

    def resolve_failure(self, failure_id: str) -> bool:
        """Mark a failure as resolved."""
        for f in self._failures:
            if f.id == failure_id:
                f.resolved = True
                self._emit("failure.resolved", {"failure_id": failure_id})
                return True
        return False

    def get_unresolved_failures(self) -> list[Failure]:
        """Get all unresolved failures."""
        return [f for f in self._failures if not f.resolved]

    def get_failures_by_category(self, category: FailureCategory) -> list[Failure]:
        """Get failures by category."""
        return [f for f in self._failures if f.category == category]

    def get_failures_by_action(self, action: str) -> list[Failure]:
        """Get failures for a specific action/tool."""
        return [f for f in self._failures if f.action == action]

    def get_recent_failures(self, count: int = 10) -> list[Failure]:
        """Get the most recent failures."""
        return self._failures[-count:]

    def get_failure_context(
        self,
        max_failures: int = 5,
        include_resolved: bool = False,
        include_stack_traces: bool = False,
    ) -> str:
        """Get formatted failure context for LLM injection."""
        failures = (
            self._failures if include_resolved
            else [f for f in self._failures if not f.resolved]
        )

        if not failures:
            return ""

        return format_failure_context(
            failures[-max_failures:],
            include_stack_traces=include_stack_traces,
        )

    def has_recent_failure(self, action: str, within_ms: float = 60000) -> bool:
        """Check if a recent failure exists for an action."""
        cutoff = time.time() - (within_ms / 1000)
        return any(
            f.action == action and f.timestamp >= cutoff
            for f in self._failures
        )

    def get_stats(self) -> dict[str, Any]:
        """Get failure statistics."""
        by_category: dict[str, int] = {}
        action_counts: dict[str, int] = {}

        for f in self._failures:
            by_category[f.category] = by_category.get(f.category, 0) + 1
            action_counts[f.action] = action_counts.get(f.action, 0) + 1

        # Top 5 most-failed actions
        most_failed = sorted(action_counts.items(), key=lambda x: x[1], reverse=True)[:5]

        return {
            "total": len(self._failures),
            "unresolved": sum(1 for f in self._failures if not f.resolved),
            "by_category": by_category,
            "most_failed_actions": most_failed,
        }

    def clear(self) -> None:
        """Clear all failures."""
        self._failures.clear()

    def on(self, listener: FailureEventListener) -> Callable[[], None]:
        """Subscribe to failure events."""
        self._listeners.append(listener)

        def unsubscribe() -> None:
            try:
                self._listeners.remove(listener)
            except ValueError:
                pass

        return unsubscribe

    def _detect_patterns(self) -> None:
        """Detect failure patterns."""
        unresolved = [f for f in self._failures if not f.resolved]

        # Repeated action pattern
        action_groups: dict[str, list[Failure]] = {}
        for f in unresolved:
            if f.action not in action_groups:
                action_groups[f.action] = []
            action_groups[f.action].append(f)

        for action, failures in action_groups.items():
            if len(failures) >= 3:
                pattern = FailurePattern(
                    type="repeated_action",
                    description=f"'{action}' has failed {len(failures)} times",
                    failure_ids=[f.id for f in failures],
                    confidence=min(0.9, 0.3 + len(failures) * 0.1),
                    suggestion=f"Consider a different approach than '{action}'",
                )
                self._emit("pattern.detected", {"pattern": pattern})

        # Category cluster pattern
        recent = self._failures[-10:] if len(self._failures) >= 10 else self._failures
        cat_counts: dict[str, list[str]] = {}
        for f in recent:
            if f.category not in cat_counts:
                cat_counts[f.category] = []
            cat_counts[f.category].append(f.id)

        for cat, ids in cat_counts.items():
            if len(ids) >= 5:
                pattern = FailurePattern(
                    type="category_cluster",
                    description=f"{len(ids)} recent '{cat}' failures",
                    failure_ids=ids,
                    confidence=0.7,
                    suggestion=f"Systematic '{cat}' issue detected - investigate root cause",
                )
                self._emit("pattern.detected", {"pattern": pattern})

    def _emit(self, event: str, data: dict[str, Any]) -> None:
        for listener in self._listeners:
            try:
                listener(event, data)
            except Exception:
                pass


def categorize_error(error: str) -> FailureCategory:
    """Categorize an error string."""
    for pattern, category in _CATEGORY_PATTERNS:
        if pattern.search(error):
            return category
    return FailureCategory.UNKNOWN


def generate_suggestion(failure: Failure) -> str:
    """Generate a suggestion based on failure category."""
    suggestions: dict[FailureCategory, str] = {
        FailureCategory.PERMISSION: "Check file permissions or use an allowed alternative.",
        FailureCategory.NOT_FOUND: "Verify the path exists. Use glob/search to find the correct location.",
        FailureCategory.SYNTAX: "Check syntax carefully. Review quoting and escaping.",
        FailureCategory.TYPE: "Check argument types match the expected schema.",
        FailureCategory.NETWORK: "Check network connectivity. Retry with backoff.",
        FailureCategory.TIMEOUT: "Try a simpler command or reduce scope.",
        FailureCategory.VALIDATION: "Check required fields and value constraints.",
        FailureCategory.RESOURCE: "Reduce memory/disk usage or free resources.",
        FailureCategory.RUNTIME: "Review the logic and check for edge cases.",
        FailureCategory.LOGIC: "Re-examine the approach and assumptions.",
    }
    return suggestions.get(failure.category, "Review the error and try a different approach.")


def format_failure_context(
    failures: list[Failure],
    include_stack_traces: bool = False,
) -> str:
    """Format failures for LLM context injection."""
    if not failures:
        return ""

    lines = ["[Previous Failures - Learn from these]"]
    for f in failures:
        status = "resolved" if f.resolved else "unresolved"
        lines.append(f"- [{f.category}] {f.action}: {f.error[:100]} ({status})")
        if f.suggestion:
            lines.append(f"  Suggestion: {f.suggestion}")
        if include_stack_traces and f.stack_trace:
            lines.append(f"  Stack: {f.stack_trace[:200]}")
    return "\n".join(lines)


def create_repeat_warning(action: str, count: int, suggestion: str | None = None) -> str:
    """Create a warning message for repeated failures."""
    msg = f"Warning: '{action}' has failed {count} times."
    if suggestion:
        msg += f" {suggestion}"
    return msg
